request-cancel-external-workflow/workflow_test.go

```diff
@@ -0,0 +1,183 @@
+			c.Receive(ctx, &signalValue)
+			logger.Info("signal received", "signal", signalName, signalValue)
+		})
+		sel.Select(ctx)
+	})
+
+	if err := workflow.Await(ctx, func() bool {
+		return signalValue == signalValueDone
+	}); err != nil {
+		logger.Error("await err", "err", err)
+	}
+	logger.Info("released from block")
+	return nil
+}
+
+// https://docs.temporal.io/docs/go/testing/
+type CancelTestSuite struct {
+	suite.Suite
+	testsuite.WorkflowTestSuite
+	env *testsuite.TestWorkflowEnvironment
+}
+
+// SetupSuite https://pkg.go.dev/github.com/stretchr/testify/suite#SetupAllSuite
+func (s *CancelTestSuite) SetupSuite() {
+
+}
+
+// SetupTest https://pkg.go.dev/github.com/stretchr/testify/suite#SetupTestSuite
+// CAREFUL not to put this `env` inside the SetupSuite or else you will
+// get interleaved test times between parallel tests (testify runs suite tests in parallel)
+func (s *CancelTestSuite) SetupTest() {
+	s.env = s.NewTestWorkflowEnvironment()
+}
+
+// BeforeTest https://pkg.go.dev/github.com/stretchr/testify/suite#BeforeTest
+func (s *CancelTestSuite) BeforeTest(suiteName, testName string) {
+
+}
+
+// AfterTest https://pkg.go.dev/github.com/stretchr/testify/suite#AfterTest
+func (s *CancelTestSuite) AfterTest(suiteName, testName string) {
+	s.env.AssertExpectations(s.T())
+}
+
+func (s *CancelTestSuite) Test_Cancel_NoMocky() {
+
+	s.env.RegisterWorkflow(MyWorkflow)
+	s.env.RegisterWorkflow(MyChild)
+	//var cancelRequestCalled bool
+	var childCancelInfo *workflow.Info
+	var childCompletedInfo *workflow.Info
+	s.env.SetOnChildWorkflowCanceledListener(func(workflowInfo *workflow.Info) {
+		childCancelInfo = workflowInfo
+	})
+	s.env.SetOnChildWorkflowCompletedListener(func(workflowInfo *workflow.Info, result converter.EncodedValue, err error) {
+		childCompletedInfo = workflowInfo
+	})
+	//s.env.OnRequestCancelExternalWorkflow(mock.Anything, childWorkflowID, "").Run(func(args mock.Arguments) {
+	//	cancelRequestCalled = true
+	//}).Return(nil).Once()
+
+	s.env.RegisterDelayedCallback(func() {
+		s.env.SignalWorkflowByID("default-test-workflow-id", signalName, signalValueDone)
+	}, time.Second*1)
+	s.env.ExecuteWorkflow(MyWorkflow)
+	s.True(s.env.IsWorkflowCompleted())
+	s.NoError(s.env.GetWorkflowError())
+	//s.True(cancelRequestCalled)
+	s.NotNil(childCancelInfo)
+	s.NotNil(childCompletedInfo)
+
+}
+func (s *CancelTestSuite) Test_Cancel_WithMocky() {
+
+	s.env.RegisterWorkflow(MyWorkflow)
+	s.env.RegisterWorkflow(MyChild)
+	//var cancelRequestCalled bool
+	var childCancelInfo *workflow.Info
+	var childCompletedInfo *workflow.Info
+	// this mock is broken...MyChild does not require any arguments but the workflowInterceptor in our test suite panics without args
+	s.env.OnWorkflow(MyChild, mock.Anything).Return(func(ctx workflow.Context) error {
+		return nil
+	})
+	s.env.SetOnChildWorkflowCanceledListener(func(workflowInfo *workflow.Info) {
+		childCancelInfo = workflowInfo
```

Comment:
This does not get called, presumably because we are mocking the ChildWorkflow...but note that the `SetOnChildWorkflowCompletedListener` _does_ get called, so it is a little surprising (not obvious).

Reply:
Looking at the go test suite code I think the current behavior is correct. Right now the workflow assumes the child WF will be blocked until it is signaled so when you call `RequestCancelExternalWorkflow` you know it will cancel the child WF. In this test since you mock the child WF, the child WF will instantly complete so your call to `RequestCancelExternalWorkflow` will not cancel the child WF since it is already complete (not running) so you don't get the callback.